Feature/#12 분과 엔티티 생성 및 crud api 구현

myeowon opened the pull request:
## 🔥 연관된 이슈

close: #12 

## 📜 작업 내용

- 분과 엔티티를 생성했습니다.
- 팀 엔티티에 분과id 필드를 추가했습니다.
- 분과 CRUD API를 개발했습니다.
  - 분과 등록
  - 해당 대회의 분과 전체 조회
  - 분과 수정
  - 분과 삭제

## 💬 리뷰 요구사항

- 리뷰 요구사항을 작성해주세요.

## ✨ 기타

- 오늘의 한마디

myeowon committed:
feat : 분과 엔티티 생성

myeowon committed:
feat : 분과 컨트롤러

myeowon committed:
feat : 분과 수정

myeowon committed:
feat : 분과 삭제

myeowon committed:
feat : 분과 검증 시 대회를 검증하므로 대회 검증 코드 삭제

myeowon committed:
feat : 해당 대회의 분과 조회

pykido reviewed (approved):
수고하셨습니다!

JJimini commented on src/main/java/com/opus/opus/modules/contest/domain/ContestTrack.java:
```diff
@@ -0,0 +1,46 @@
+    private String trackName;
+
+    @Column(nullable = false)
+    private Long contestId;
```
`ContestTrack`은 `Contest` module 안에 들어있기 때문에 식별자가 아니라 Mapping 걸어주시면 됩니다!

myeowon replied:
넵 알겠습니다!

JJimini commented on src/main/java/com/opus/opus/modules/contest/api/ContestTrackController.java:
```diff
@@ -0,0 +1,64 @@
+    @Secured("ROLE_관리자")
+    public ResponseEntity<Void> createContestTrack(
+            @Valid @RequestBody final ContestTrackRequest request,
+            @PathVariable final Long contestId) {
```
이거 포맷팅이 깨진 것 같아요~!

JJimini commented on src/main/java/com/opus/opus/modules/contest/application/ContestTrackCommandService.java:
```diff
@@ -0,0 +1,44 @@
+
+@Service
+@Transactional
+@RequiredArgsConstructor
+public class ContestTrackCommandService {
+    private final ContestTrackRepository contestTrackRepository;
```
nit: 개행해주세요..!

JJimini commented on src/main/java/com/opus/opus/modules/contest/domain/dao/ContestTrackRepository.java:
```diff
@@ -0,0 +1,14 @@
+public interface ContestTrackRepository extends JpaRepository<ContestTrack, Long> {
+    boolean existsByContestIdAndTrackName(final Long contestId, final String trackName);
+
+    Optional<ContestTrack> findByIdAndContestId(final Long trackId, final Long contestId);
+
```
👍

JJimini commented on src/main/java/com/opus/opus/modules/team/application/convenience/TeamConvenience.java:
```diff
@@ -0,0 +1,29 @@
+            throw new TeamException(CONTEST_HAS_TEAM);
+        }
+    }
+
+    public void validateAllTeamsDeletedInTrack(final Long trackId) {
```
앞에 category pr의 삭제 확인 메서드와 네이밍 규칙이 좀 다른 것 같아요! 통일하면 좋을 것 같습니다~

myeowon replied:
대회를 삭제할 때 속한 팀이 있는지 확인하는 함수로 validateAllTeamsDeletedInContest
분과를 삭제할 때 속한 팀이 있는지 확인하는 함수로 validateAllTeamsDeletedInTrack을 사용하는데
네이밍을 어떻게 하는게 좋을까요??

JJimini replied:
앞 pr에 카테고리 삭제할 때 속한 대회가 있는지 확인하는 함수로 `validateAllContestsDeleted`를 사용하셨어서 In** 이 부분이 통일되면 더 좋을 것 같습니다! 예를 들어서 `validateAllContestsDeleted`도 In**를 뒤에 붙이면 더 좋을 것 같아요!

myeowon replied:
넵 카테고리 PR에 수정해서 올리겠습니다!

JJimini commented on src/main/java/com/opus/opus/modules/team/domain/Team.java:
```diff
@@ -53,6 +53,9 @@ public class Team extends BaseEntity {
     private Long contestId;
 
+    @Column(nullable = false)
+    private Long trackId;
```
모든 팀은 track을 가지나요??

myeowon replied:
없는 경우도 있을 거 같아서 삭제했습니다!

JJimini reviewed (commented):
수고하셨습니다~ entity 관련 comment 확인해주세요!!

myeowon committed:
fix : trackId nullable로 변경

JJimini reviewed (approved):
네이밍 관련 내용 말고(네이밍도 카테고리 관련임..)는 수정이 완료된 것 같아서 approve하겠습니다~!

myeowon committed:
feat : 분과 등록

# Conflicts:
#	src/main/java/com/opus/opus/modules/contest/application/convenience/ContestConvenience.java
#	src/main/java/com/opus/opus/modules/contest/exception/ContestExceptionType.java
